### DISCUSSION TITLE
Tournament selection and optimisation

### DISCUSSION THREAD

**Discussion by Flowake** [ahmedfgad/GeneticAlgorithmPython#54] [Ideas]
Hello and thanks for this great python library,

I was going through the code to understand how some parts of the project work and I found a way to optimise the tournament parent selection so that it doesn't compute uncesseray fitness values.

In my case, fitness evaluation can take a lot of time (up to 30s for a solution in some cases) and for some parent selection technique the fitness of the whole population is required.

But for the tournament selection technique, only the fitnesses of the selected individuals need to be computed. For example if I only select 20% of the solutions to be parents of the next generation, and I use a 3 way tournament, I will only need up to 60% of the whole population fitness.

It might require a little bit of redesign of the run function, but it might be possible to skip the full fitness computation and wait for the parent selection to compute the selected solution fitness and store them if they are once again selected to compete.

**Comment by ahmedfgad** [comment]
Thanks for using PyGAD!

This is indeed a useful request. I adjusted the fitness function to reuse the already calculated parent fitness values rather than calling the fitness function. Check this commit: https://github.com/ahmedfgad/GeneticAlgorithmPython/commit/c87641bb9f774cebc40a45e70834832b04ae32b5

Please let me know if you have any suggestions.
